### PR TITLE
potential fix to buildRows crash caused by switching selected tab

### PR DESCRIPTION
**Pull request by chassu**
Deferring the selectedTab update until after tabsUpdatedPublisher.send() is finished to make sure that - when buildRows() is invoked, it's not using an old version of allDetails. The change is based on discussion with @darinwf. Conceptually, I hope this is the correct fix. PTAL.

### How was this tested?
I don't know a reliable way to repro the bug. Is there a way to repro it?

### CODE REVIEW

File: Client/Frontend/Browser/Tabs/TabManager/TabManager.swift [hunk 1]
```diff
@@ -34,6 +34,10 @@ class TabManager: NSObject {
     var selectedTab: Tab?
     var selectedTabPublisher = PassthroughSubject<Tab?, Never>()
 
+    // Variables used to send a selectedTab change signal
+    var previousTab: Tab?
```
Comment (darinwf) [inline Client/Frontend/Browser/Tabs/TabManager/TabManager.swift:38]: We shouldn't need to use a member variable to communicate this state. You can just sample `selectedTab` inside `removeTabs` before it is updated.

File: Client/Frontend/Browser/Tabs/TabManager/TabManager+CloseTabs.swift [hunk 2]
```diff
@@ -43,6 +43,27 @@ extension TabManager {
...
+            selectedTabPublisher.send(tab)
+        }
+
+        if let tab = previousTab {
+            TabEvent.post(.didLoseFocus, for: tab)
+        }
+
+        if let tab = selectedTab {
+            TabEvent.post(.didGainFocus, for: tab)
+            tab.applyTheme()
```
[the remark this answers is not in the document]
Comment (darinwf) [reply Client/Frontend/Browser/Tabs/TabManager/TabManager+CloseTabs.swift:64]: To be honest, this code probably doesn't belong in TabManager at all, but we can deal with that issue later.

**Review by darinwf — APPROVED** [review]
LGTM!

**Comment by chassu** [comment]
I'm not sure if I have a valid question here - but I realized that the logic applying special treatment to selectedTab's border is not governed by selectedTabNotifications(). I think `selectedTab = tab` in selectTab() simply achieves this, because the definition of CardDetails.isSelected is    

```
var isSelected: Bool {
        self.manager.selectedTab?.tabUUID == id
    }

```

I've always thought the most important part of selectTab() is to highlight the new selected tab, but with the way I understand it now, it seems like this code block:

```
        delegates.forEach {
            $0.get()?.tabManager(
                self, didSelectedTabChange: selectedTab, previous: previous,
                isRestoring: store.isRestoringTabs,
                updateZeroQuery: true)
        }

        if let tab = selectedTab {
            selectedTabPublisher.send(tab)
        }
```

has nothing to do with it.

**Comment by chassu** [comment]
I think my question is kind of like "besides highlighting the border of a different card, what other changes are really important enough for tabManager to signal a selectedTab change?"

**Comment by chassu** [comment]
^I don't know if I have a point here, but I had this thought because I changed `notify:true` to false in removeTab() and surprised that a new card is still getting selected.

```
    func removeTab(_ tab: Tab, showToast: Bool = false, updateSelectedTab: Bool = true) {
        guard let index = tabs.firstIndex(where: { $0 === tab }) else { return }
        addTabsToRecentlyClosed([tab], showToast: showToast)
        removeTab(tab, flushToDisk: true, notify: true)

        if updateSelectedTab {
            updateSelectedTabAfterRemovalOf(tab, deletedIndex: index, notify: false)
        }
    }
```

**Comment by darinwf** [comment]
There are probably redundant updates happening somehow.

**Pull request merged**